src/vmm/src/devices/virtio/net/metrics.rs

```diff
@@ -0,0 +1,308 @@
+//!     ...
+//!  }
+//!  "net1": {
+//!     "activate_fails": "SharedIncMetric",
+//!     "cfg_fails": "SharedIncMetric",
+//!     "mac_address_updates": "SharedIncMetric",
+//!     "no_rx_avail_buffer": "SharedIncMetric",
+//!     "no_tx_avail_buffer": "SharedIncMetric",
+//!     ...
+//!  }
+//!  ...
+//!  "netN": {
+//!     "activate_fails": "SharedIncMetric",
+//!     "cfg_fails": "SharedIncMetric",
+//!     "mac_address_updates": "SharedIncMetric",
+//!     "no_rx_avail_buffer": "SharedIncMetric",
+//!     "no_tx_avail_buffer": "SharedIncMetric",
+//!     ...
+//!  }
+//! }
+//! ```
+//! Each `net` field in the example above is a serializable `NetDeviceMetrics` structure
+//! collecting metrics such as `activate_fails`, `cfg_fails`, etc. for the network device.
+//! `net0`, `net1` and `netN` in the above example represent metrics 0th, 1st and 'N'th
+//! network device respectively and `net` is the aggregate of all the per device metrics.
+//!
+//! # Limitations
+//! Network device currently do not have `vmm::logger::metrics::StoreMetrics` so aggregate
+//! doesn't consider them.
+//!
+//! # Design
+//! The main design goals of this system are:
+//! * To improve network device metrics by logging them at per device granularity.
+//! * Continue to provide aggregate net metrics to maintain backward compatibility.
+//! * Move NetDeviceMetrics out of from logger and decouple it.
+//! * Use lockless operations, preferably ones that don't require anything other than simple
+//!   reads/writes being atomic.
+//! * Rely on `serde` to provide the actual serialization for writing the metrics.
+//! * Since all metrics start at 0, we implement the `Default` trait via derive for all of them, to
+//!   avoid having to initialize everything by hand.
+//!
+//! The system implements 1 types of metrics:
+//! * Shared Incremental Metrics (SharedIncMetrics) - dedicated for the metrics which need a counter
+//! (i.e the number of times an API request failed). These metrics are reset upon flush.
+//! We use NET_DEV_METRICS_PVT instead of adding an entry of NetDeviceMetrics
+//! in Net so that metrics are accessible to be flushed even from signal handlers.
+
+use serde::ser::SerializeMap;
+use serde::{Serialize, Serializer};
+
+use crate::logger::{IncMetric, SharedIncMetric};
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct NetDeviceMetricsIndex(usize);
+
+impl NetDeviceMetricsIndex {
+    pub fn get(&self) -> &NetDeviceMetrics {
+        // SAFETY: This is called by individual Net devices
+        // on self (i.e. self.metrics.get()) so no race condition
+        // and it is safe.
+        unsafe { &NET_DEV_METRICS_PVT.metrics[self.0] }
+    }
+}
+
+/// provides instance for net metrics
+#[derive(Debug)]
+pub struct NetDeviceMetricsPool {
+    /// used to access per net device metrics
+    metrics: Vec<NetDeviceMetrics>,
+}
+
+impl NetDeviceMetricsPool {
+    /// default construction
+    pub fn get() -> NetDeviceMetricsIndex {
```

Comment:
Maybe the name `new_index` or `allocate` is better than `get`.

Reply:
done